cmd/dfdaemon/main.go

```diff
@@ -15,40 +15,97 @@
 	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
 	"runtime"
 
-	"github.com/sirupsen/logrus"
-
 	"github.com/dragonflyoss/Dragonfly/cmd/dfdaemon/options"
 	"github.com/dragonflyoss/Dragonfly/dfdaemon/initializer"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+var (
+	opt           = &options.Options{}
+	defaultDFRepo = filepath.Join(os.Getenv("HOME"), ".small-dragonfly/dfdaemon/data/")
 )
 
 func main() {
-	options := options.NewOption()
-	options.AddFlags(flag.CommandLine)
-	flag.Parse()
+	var rootCmd = &cobra.Command{
+		Use:   "dfdaemon",
+		Short: "The dfdaemon is a proxy that intercepts image download requests.",
+		Long:  "The dfdaemon is a proxy between container engine and registry used for pulling images.",
+		Run: func(cmd *cobra.Command, args []string) {
+			runDaemon()
+		},
+	}
 
-	initializer.Init(options)
+	setupFlags(rootCmd, opt)
 
-	// if CommandLine.MaxProcs <= 0, programs run with GOMAXPROCS set to the number of cores available
-	if options.MaxProcs > 0 {
-		runtime.GOMAXPROCS(options.MaxProcs)
+	if err := rootCmd.Execute(); err != nil {
+		logrus.Error(err)
+		os.Exit(1)
 	}
+}
 
-	logrus.Infof("start dfdaemon param:%+v", options)
+// setupFlags setups flags for command line.
+func setupFlags(cmd *cobra.Command, o *options.Options) {
+	fs := cmd.Flags()
 
-	fmt.Printf("\nlaunch dfdaemon on port:%d\n", options.Port)
+	fs.BoolVar(&o.Help, "h", false, "help")
+	fs.BoolVar(&o.Version, "v", false, "version")
+	fs.BoolVar(&o.Verbose, "verbose", false, "verbose")
+	fs.IntVar(&o.MaxProcs, "maxprocs", 4, "the maximum number of CPUs that the dfdaemon can use")
 
-	var err error
+	// http server config
+	fs.StringVar(&o.HostIP, "hostIp", "127.0.0.1", "dfdaemon host ip, default: 127.0.0.1")
+	fs.UintVar(&o.Port, "port", 65001, "dfdaemon will listen the port")
+	fs.StringVar(&o.CertFile, "certpem", "", "cert.pem file path")
+	fs.StringVar(&o.KeyFile, "keypem", "", "key.pem file path")
 
-	if options.CertFile != "" && options.KeyFile != "" {
-		err = http.ListenAndServeTLS(fmt.Sprintf(":%d", options.Port),
-			options.CertFile, options.KeyFile, nil)
+	// dfget download config
+	fs.StringVar(&o.DFRepo, "localrepo", defaultDFRepo, "temp output dir of dfdaemon")
+	fs.StringVar(&o.CallSystem, "callsystem", "com_ops_dragonfly", "caller name")
+	fs.StringVar(&o.DfPath, "dfpath", "", "dfget path")
+	fs.StringVar(&o.RateLimit, "ratelimit", "", "net speed limit,format:xxxM/K")
+	fs.StringVar(&o.URLFilter, "urlfilter", "Signature&Expires&OSSAccessKeyId", "filter specified url fields")
+	fs.StringVar(&o.Registry, "registry", "https://index.docker.io", "registry addr(https://abc.xx.x or http://abc.xx.x) and must exist if dfdaemon is used to mirror mode")
+	fs.StringVar(&o.DownRule, "rule", "", "download the url by P2P if url matches the specified pattern,format:reg1,reg2,reg3")
+	fs.BoolVar(&o.Notbs, "notbs", true, "not try back source to download if throw exception")
+}
+
+// start to run dfdaemon server.
+func runDaemon() {
```

Comment:
I think that we should move the operation about initing config into the single function and process it ahead of the runDaemon.